Tools/build/check_warnings.py

```diff
@@ -2,39 +2,88 @@
 Parses compiler output with -fdiagnostics-format=json and checks that warnings
 exist only in files that are expected to have warnings.
 """
+
 import argparse
+from collections import defaultdict
 import json
 import re
 import sys
 from pathlib import Path
 
 
-def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
+def extract_warnings_from_compiler_output_clang(
+    compiler_output: str,
+) -> list[dict]:
     """
-    Extracts warnings from the compiler output when using
-    -fdiagnostics-format=json
+    Extracts warnings from the compiler output when using clang
+    """
+    # Regex to find warnings in the compiler output
+    clang_warning_regex = re.compile(
+        r"(?P<file>.*):(?P<line>\d+):(?P<column>\d+): warning: (?P<message>.*)"
+    )
+    compiler_warnings = []
+    for line in compiler_output.splitlines():
+        if match := clang_warning_regex.match(line):
+            compiler_warnings.append(
+                {
+                    "file": match.group("file"),
+                    "line": match.group("line"),
+                    "column": match.group("column"),
+                    "message": match.group("message"),
+                }
+            )
 
-    Compiler output as a whole is not a valid json document, but includes many
-    json objects and may include other output that is not json.
+    return compiler_warnings
+
+
+def extract_warnings_from_compiler_output_json(
+    compiler_output: str,
+) -> list[dict]:
     """
+    Extracts warnings from the compiler output when using
+    -fdiagnostics-format=json.
 
+    Compiler output as a whole is not a valid json document,
+    but includes many json objects and may include other output
+    that is not json.
+    """
     # Regex to find json arrays at the top level of the file
     # in the compiler output
     json_arrays = re.findall(
-        r"\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*\])*\])*\]", compiler_output
+        r"\[(?:[^[\]]|\[[^\]]*\])*\]", compiler_output
     )
     compiler_warnings = []
     for array in json_arrays:
         try:
             json_data = json.loads(array)
             json_objects_in_array = [entry for entry in json_data]
-            compiler_warnings.extend(
-                [
-                    entry
-                    for entry in json_objects_in_array
-                    if entry.get("kind") == "warning"
-                ]
-            )
+            warning_list = [
```

Comment:
This PR depends on https://github.com/python/cpython/pull/122211 and contains code that has been previously reviewed there. Should not merge prior to that PR

Reply:
https://github.com/python/cpython/pull/122211 has been merged, please can you fix the conflicts here?

Reply:
Resolved conflicts with main